// @coreui/vue/dist/cjs/components/tabs/CTabs.js
'use strict';

var vue = require('vue');

const CTabs = vue.defineComponent({
    name: 'CTabs',
    props: {
        /**
         * The active item key.
         */
        activeItemKey: {
            type: [Number, String],
            required: true,
        }
    },
    emits: [
        /**
         * The callback is fired when the active tab changes.
         */
        'change',
    ],
    setup(props, { slots, emit }) {
        const activeItemKey = vue.ref(props.activeItemKey);
        const uniqueId = vue.useId();
        const setActiveItemKey = (key) => {
            activeItemKey.value = key;
        };
        vue.watch(() => props.activeItemKey, (value) => {
            activeItemKey.value = value;
            emit('change', value);
        });
        vue.provide('activeItemKey', activeItemKey);
        vue.provide('id', uniqueId);
        vue.provide('setActiveItemKey', setActiveItemKey);
        return () => vue.h('div', { class: 'tabs' }, slots.default && slots.default());
    },
});

exports.CTabs = CTabs;
